modules/auxiliary/gather/memcached_extractor.rb

```diff
@@ -58,8 +58,8 @@ def enumerate_keys
         sock.send("stats cachedump #{sid} #{max_keys}\r\n", 0)
         data = sock.recv(4096)
         break if !data || data.length == 0
-        matches = /^ITEM (?<key>.*) \[/.match(data)
-        keys << matches[:key] if matches
+        matches = data.scan(/^ITEM (?<key>.*) \[/)
+        keys = keys + matches.flatten! if matches
```

Comment:
I haven't test this, but it looks like this rewrite might have the same issue as the below changes? i.e. Raising an error on nil due to the `matches.flatten!` call.

I wonder if the previous implementation was a cleaner approach:

```ruby
        matches = /^ITEM (?<key>.*) \[/.match(data)
        keys << matches[:key] if matches
```

Was there a bug in the existing implementation that was being fixed?

Reply:
Using cachedump we collect a list of slab_ids before enumerating keys. When there are no keys there are no slabs (as far as I have seen) so this code doesn't run. I'll add the extra check in as done in enumerate_keys_lru just to be sure. 

The original implementation was only ever returning the first key in my tests, even on older versions of memcached. This is my first foray into Ruby but it looks like that's how .match is implemented, so I do believe this was an additional bug.